### ISSUE TITLE
`pipeline.pipeline` confuses IDE autocompletion

### ISSUE THREAD

**Issue by noklam** [kedro-org/kedro#2805]
## Description

We import `pipeline` in our starter and this confuse IDE like VSCode, as it thinks `pipeline` is a module instead of the `pipeline` function.

`from kedro.pipeline import Pipeline, node, pipeline`

Changing it to the following avoid the issue.

`from kedro.pipeline.modular_pipeline import pipeline`
![image](https://github.com/kedro-org/kedro/assets/18221871/6430a53f-118e-430d-bf7c-cb3092c8ddad)

The problem here is IDE confuse `kedro.pipeline.modular_pipeline` as `kedro.pipeline.pipeline` module.

## Context

Solution 1:
- Move the `pipeline` definition (Or what's the difference between pipeline and modular pipeline? Will be a breaking change if we move modules)

Solution 2:
- Change starter to import from `modular_pipeline` explicitly, should we keep the shortcut in this case?

Solution 3:
- find out ways that won't confuse IDE (VS Code in my case, user report this with PyCharm too)

## Actual Result

```
-- If you received an error, place it here.
```

```
-- Separate them if you have more than one.
```

## Your Environment

* Kedro version used (`pip show kedro` or `kedro -V`):
* Python version used (`python -V`):
* Operating system and version:

**Comment by astrojuanlu** [comment]
or

```
from kedro import Pipeline

Pipeline.make(...)
```

😄 #712

(see also #2723 about "modular pipelines" vs just "pipelines")

**Comment by noklam** [comment]
Discussed in backlog grooming. Will keep this as a low priority for the 0.19 release and see if we have time to fit this in.

If we end up don't want to break anything in 0.19, we should still update the starters to avoid these syntax highlight problem by importing from `kedro.pipeline.modular_pipeline` directly.

**Comment by astrojuanlu** [comment]
There's more confusion: https://github.com/kedro-org/kedro-viz/issues/1522#issue-1885805761

Some users are creating the pipelines with the `Pipeline` class.

I think we should tackle this and #2723 at one.

**Comment by astrojuanlu** [comment]
I just saw this code snippet in someone else's code:

```python
my_pipeline = Pipeline([
    Node(filter_func, "df_input", "df_filtered"),
    Node(actual_func, "df_filtered", "df_output"),
])
```

is there a reason why we are using the `pipeline` and `node` helpers instead of the `Pipeline` and `Node` class initializers directly?

**Comment by rashidakanchwala** [comment]
was also wondering if we could do namespace as part of Pipeline? 
```
my_pipeline = Pipeline([
    Node(filter_func, "df_input", "df_filtered"),
    Node(actual_func, "df_filtered", "df_output"),
    namespace = 'something'
])
```

**Comment by rashidakanchwala** [comment]
yes, i put it as we are investigating this issue on Polishing API surface. But i just wanted to confirm a few things. so i will put it with more context

**Comment by merelcht** [comment]
Yeah I have been thinking along the same lines. I find it hard to say if this is a good or bad idea without playing around with it all a bit more though. I've started on #4507, which will help massively in all this.

**Comment by lamhm** [comment]
@rashidakanchwala 
> Would it be possible to remove `modular_pipeline` entirely and instead introduce `namespace` as an optional parameter in the `Pipeline` constructor?

I assume this idea means to deprecate the `pipeline` function so users can only create a `Pipeline` via its constructor. I like this idea even though it would mean I will have to revise much of my code. It bugged me (and confused me) when I found out that `pipeline()` and `Pipeline()` do almost the same thing but slightly differently.

So far, my team have accepted that Kedro is still experimental and that we may have to spend a few days fixing compatibility issues every time we upgrade Kedro on our platform (because of this, we only upgrade Kedro once a year). But we hope reaching version 1.0 means reaching stability. I think the Kedro team should take this chance to prioritise long-term stability over backward compatibility (speaking as a long-time user).

**Comment by merelcht** [comment]
@lamhm The POC in https://github.com/kedro-org/kedro/pull/4563 demonstrates what we want to do in a non-breaking way, but eventually the whole `modular_pipeline` module will be removed.

**Comment by astrojuanlu** [comment]
All sub-issues here have been closed 🎉
